File: lua/queen-attack/queen-attack.lua
local Queen = {}
Queen.__index = Queen

function Queen:new(pos)
   if pos.row < 0 or pos.row > 7 or pos.column < 0 or pos.column > 7 then
      error 'invalid coordinates'
   end
   local newQueen = {row=pos.row, column=pos.column}
   newQueen.can_attack = function (other)
      return newQueen:canAttack(other)
   end
   return setmetatable(newQueen, self)
end

function Queen:__tostring()
   return "Q["..self.row..","..self.column.."]"
end

function Queen:canAttack(other)
   return self.column == other.column
      or self.row == other.row
      or math.abs(self.row - other.row) == math.abs(self.column - other.column)
end


return function (pos) return Queen:new(pos) end

File: python/queen-attack/queen_attack.py
# pylint: disable=too-few-public-methods
class Queen:
    def __init__(self, row, column):
        self.row = row
        self.col = column
        self._validate()

    def _validate(self):
        if self.row > 7:
            raise ValueError("row not on board")
        if self.row < 0:
            raise ValueError("row not positive")
        if self.col > 7:
            raise ValueError("column not on board")
        if self.col < 0:
            raise ValueError("column not positive")

    def can_attack(self, other):
        if self.row == other.row and self.col == other.col:
            raise ValueError("Invalid queen position: both queens in the same square")
        return (
            self.row == other.row
            or self.col == other.col
            or abs(self.col - other.col) == abs(self.row - other.row)
        )
